mbed_build/_internal/config_header_file.py

```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (C) 2020 Arm Mbed. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Module in charge of mbed_config.h file generation."""
+import pathlib
+
+import jinja2
+
+from typing import List
+from mbed_build._internal.config.config import Config, Option, Macro
+from mbed_build._internal.config.assemble_build_config import assemble_config
+
+TEMPLATES_DIRECTORY = pathlib.Path("_internal", "templates")
+TEMPLATE_NAME = "mbed_config.tmpl"
+
+
+def generate_config_header_file(mbed_target: str, program_path: str) -> str:
+    """Generate the program's mbed_config.h file containing the configuration definitions.
+
+    Args:
+        mbed_target: the target the application is being built for
+        program_path: the path to the local Mbed program
+
+    Returns:
+        A string of rendered contents for the file.
+    """
+    config = assemble_config(mbed_target, pathlib.Path(program_path))
+    return _render_config_header_template(config)
+
+
+def _render_config_header_template(config: Config) -> str:
+    """Renders the mbed_config.h template with the correct config definitions.
+
+    Args:
+        config: a Config object containing the combined configuration data for the build
+
+    Returns:
+        The rendered contents of the template with the config definitions included.
+    """
+    context = {
```

Comment:
An example more generic option I have in mind:

```python
def _max_attribute_length(objects: Iterable[object], attribute: str) -> int:
    attrs = (getattr(o, attribute) for o in objects)
    try:
        return max(len(attr) for attr in attrs if attr is not None)
    except ValueError:  # no attrs found
        return 0
```

Reply:
I will make the changes you suggest for 1 and 3. 👍 

With 2 I could just use the raw form but I was hitting the problem that the `value` could be a number or a bool which wouldn't let be do the justifying I wanted to.  I can't str() it in the template, and I don't want to re-set the stringified value back onto the Option and Macro objects so I converted it into a tuple to be immutable.

Reply:
That is a lot nicer. Thanks. 
7273047 and 50905c3

Reply:
Unsure if it's better or worse, but if it's just for the sake of presentation, you could register a small filter for your jinja environment https://jinja.palletsprojects.com/en/2.11.x/api/#custom-filters.

```
def ljust(value, width):
   return str(value).ljust(width)

def _render_config_header_template(config: Config) -> str:
   ...
   env.filters['ljust'] = ljust
```

In your template:
```
{% for option in options -%}
 {% if option.value is not none -%}
 #define {{ option.name|ljust(max_name_length) }} {{ option.value|ljust(max_value_length) }} // set by {{ set_by }}
 {% endif %}
 {%- endfor %}
```

The benefit of this approach is that adding, removal or renaming of config property would need to happen in one place only. :)

Reply:
🤔 Thats true, I had wanted to avoid going down that route after flashbacks to fun with trying to get custom filters to be happy in both Jinja and Django templates on the web team, but maybe it wont be so bad here (no Django templates). I'll give it a go 👍

Reply:
edf9c5a